### PR TITLE
Feature/aes

### PR DESCRIPTION
**Pull request by cyberzac**
WIP Support for AES, Needs more test cases

### CODE REVIEW

File: crypto-javax/src/main/scala/cryptic/crypto/AES.scala [hunk 1]
```diff
@@ -1,36 +1,74 @@
...
-class AES(transformation: String) {
-  val cipher: Cipher = Cipher.getInstance(s"AES/$transformation")
-
-  implicit def encrypt(implicit key: SecretKey): Encrypt = (plainText: PlainText) => {
+  private val secureRandom = new SecureRandom()
+  private val factoryAlgorithm = "PBKDF2WithHmacSHA256"
+  private val keyAlgorithm = "AES"
+  private val keyspecIterationCount = 65536
+  private val keyspecLength = 256
+  private val saltLength = 32
```
Comment (odd) [inline crypto-javax/src/main/scala/cryptic/crypto/AES.scala]: Shouldn't these settings be configurable?

File: crypto-javax/src/main/scala/cryptic/crypto/AES.scala [hunk 2]
```diff
@@ -1,36 +1,74 @@
...
+
+  case class Salt(bytes: Array[Byte]) extends AnyVal {
+    def length: Int = bytes.length
+  }
+  case class AESPassword(value: String) extends AnyVal {
+    def bytes: Array[Byte] = value.getBytes
+  }
+
+  val cipher: Cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
```
Comment (odd) [inline crypto-javax/src/main/scala/cryptic/crypto/AES.scala]: Sholdn't this setting be configurable?

File: crypto-javax/src/main/scala/cryptic/crypto/AES.scala [hunk 3]
```diff
@@ -1,36 +1,74 @@
...
+    val key = keygen(password, salt)
     cipher.init(Cipher.ENCRYPT_MODE, key)
-    CipherText(cipher.doFinal(plainText))
+    val params = cipher.getParameters
+    val initVector = params.getParameterSpec(classOf[IvParameterSpec]).getIV
+    val saltLength = salt.length
+    val cipherText = cipher.doFinal(plainText)
+    val ivLength = initVector.length
+    val textLength = cipherText.length
+    // Encode with length prefix to allow for backwards compatibility
```
Comment (odd) [inline crypto-javax/src/main/scala/cryptic/crypto/AES.scala]: Backwards compatibility with what exactly?

File: crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala [hunk 4]
```diff
@@ -1,41 +1,30 @@
...
+package crypto
+
+import org.scalatest._
+
+class AESSpec extends FlatSpec with Matchers {
+  import AES._
+  implicit private val password: AESPassword = AESPassword("secret")
+  println(s"secretKey:$password")
+  private val plainText = PlainText("nisse")
+  val encryptFun: Encrypt = encrypt // Uses implicit key
```
Comment (odd) [inline crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala]: Is this necessary? Why can't you use `encrypt` directly?

File: crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala [hunk 5]
```diff
@@ -1,41 +1,30 @@
...
+
+  "AES" should "support encryption and decryption" in {
+    val encrypted = encryptFun(plainText)
+
+    val decryptFun: Decrypt = decrypt // Uses implicit key
```
Comment (odd) [inline crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala]: Is this necessary? Why can't you use `decrypt` directly?

File: crypto-javax/src/main/scala/cryptic/crypto/AES.scala (hunk 3 above)
Comment (cyberzac) [reply crypto-javax/src/main/scala/cryptic/crypto/AES.scala]: Run length encoding makes it easier to deserialise even if the format has changed.
An other option is to use more advanced format, e.g. protobuf.

File: crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala (hunk 4 above)
Comment (cyberzac) [reply crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala]: We can use the test cases as documentation. Is this a good or bad usage?

File: crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala (hunk 5 above)
Comment (cyberzac) [reply crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala]: Same as line 8-11

File: crypto-javax/src/main/scala/cryptic/crypto/AES.scala (hunk 1 above)
Comment (odd) [reply crypto-javax/src/main/scala/cryptic/crypto/AES.scala]: You don't need to store them to be able to decrypt using the same key configuration (i.e you see the key configuration as part of the key). By allowing them to be configurable you allow different clients to use different key configurations though.

File: crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala (hunk 4 above)
Comment (odd) [reply crypto-javax/src/test/scala/cryptic/crypto/AESSpec.scala]: Documentation is always good :) Perhaps the sample code should be done in a special paradox test case though (as is done in https://github.com/ScalaCrypto/cryptic/blob/master/crypto-test/src/test/scala/cryptic/GettingStartedSpec.scala)?

File: crypto-javax/src/main/scala/cryptic/crypto/AES.scala (hunk 1 above)
Comment (cyberzac) [reply crypto-javax/src/main/scala/cryptic/crypto/AES.scala]: You must ensure that you use the same configuration when reading a persisted serialized object. We should probably add something like akka's serialization config. 
And we don't want to do this for each encrypted instance. I'm not sure on the best path to explore. We probably need to meet over a whiteboard.

Comment (odd) [reply crypto-javax/src/main/scala/cryptic/crypto/AES.scala]: Yes, but ensuring that the correct key (with the correct configuration) is used by both the encrypting and decrypting sides is an application level responsibility. When decrypting encrypted data that has been serialized/persisted, the decrypting side needs to know both the key and the key configuration used (i.e. neither would be part of the serialized encrypted data itself).

Comment (cyberzac) [reply crypto-javax/src/main/scala/cryptic/crypto/AES.scala]: Added AESParams for the configuration of AES.

**Pull request merged**